src/lib/expression/abstract_expression.hpp

```diff
@@ -56,6 +56,11 @@ class AbstractExpression : public std::enable_shared_from_this<AbstractExpressio
   bool operator==(const AbstractExpression& other) const;
   bool operator!=(const AbstractExpression& other) const;
 
+  /**
+   * Set if expression was already replaced when caching to avoid double replacement
+   */
+  std::shared_ptr<AbstractExpression> replaced_by;
```

Comment:
when looking at it twice, this is actually never used... at least not by any testcase. as far as i understand, it should prevent that a value expression is replaced twice whenever it occurs twice in the plan. however, i don't see how this could ever happen with our current approach. two times the same litteral in the sql string should (and does) translate to two different value expressions and every rule that might replace these two expressions by only one such as the expression reduction rule is marked as not caching safe. so i think we can just drop this.

Reply:
Even if we had to make sure that no expression is replaced twice, I think that this should not happen in the class but in the algorithm. In any case - if it turns out that we do not need it, that's even better.